src/Storages/MergeTree/registerStorageMergeTree.cpp

```diff
@@ -688,9 +691,63 @@ static StoragePtr create(const StorageFactory::Arguments & args)
                 args.storage_def->ttl_table->ptr(), metadata.columns, context, metadata.primary_key, allow_suspicious_ttl);
         }
 
+        storage_settings->loadFromQuery(*args.storage_def, context, LoadingStrictnessLevel::ATTACH <= args.mode);
+
+        // updates the default storage_settings with settings specified via SETTINGS arg in a query
+        if (args.storage_def->settings)
+        {
+            if (args.mode <= LoadingStrictnessLevel::CREATE)
+                args.getLocalContext()->checkMergeTreeSettingsConstraints(initial_storage_settings, storage_settings->changes());
+            metadata.settings_changes = args.storage_def->settings->ptr();
+        }
+
         if (args.query.columns_list && args.query.columns_list->indices)
-            for (auto & index : args.query.columns_list->indices->children)
+        {
+            for (auto &index: args.query.columns_list->indices->children)
+            {
+                auto index_name = index->as<ASTIndexDeclaration>()->name;
+                if (index_name.starts_with(INDEX_MINMAX_NUMERIC_PREFIX) || index_name.starts_with(INDEX_MINMAX_STRING_PREFIX))
+                {
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Cannot add index {}: index has prefix {} or {} which are reserved for "
+                                                                "default minmax indices", index_name, INDEX_MINMAX_NUMERIC_PREFIX, INDEX_MINMAX_STRING_PREFIX);
+                }
                 metadata.secondary_indices.push_back(IndexDescription::getIndexFromAST(index, columns, context));
+            }
+
+            if ((*storage_settings)[MergeTreeSetting::enable_minmax_index_for_all_numeric_columns]
+            || (*storage_settings)[MergeTreeSetting::enable_minmax_index_for_all_string_columns])
+            {
+                for (const auto & column : metadata.columns)
+                {
+                    if (!isNumber(column.type) && !isString(column.type))
+                        continue;
+
+                    if (isNumber(column.type) && !(*storage_settings)[MergeTreeSetting::enable_minmax_index_for_all_numeric_columns])
+                        continue;
+
+                    if (isString(column.type) && !(*storage_settings)[MergeTreeSetting::enable_minmax_index_for_all_string_columns])
+                        continue;
+
+                    bool index_exists = false;
+                    for (auto &index: metadata.secondary_indices)
+                    {
+                        if (index.column_names.front() == column.name)
+                        {
+                            index_exists = true;
```

Comment:
If a column is defined to have a non-minmax secondary index (e.g. `set(1)`), then we'll not add an auto-generated min-max index here. For example

```sql
CREATE TABLE tbl
(
    `key` Int,
    `x` Int,
    INDEX idx x TYPE set(1)
)
ENGINE = MergeTree
ORDER BY key
SETTINGS enable_minmax_index_for_all_numeric_columns = 1
```

will not create a minmax index on `x` but it should.

Reply:
added a check to see if existing index is minmax, please let me know if you have any suggestions.